flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/HeaderProducer.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.kafka.sink;
+
+import org.apache.flink.annotation.PublicEvolving;
+
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.Headers;
+import org.apache.kafka.common.header.internals.RecordHeaders;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+
+/** Creates an {@link Iterable} of {@link Header}s from the input element. */
+@PublicEvolving
+public interface HeaderProducer<IN> extends Serializable {
+    default Headers produceHeaders(IN input) {
+        return new RecordHeaders(new ArrayList<>());
```

Comment:
Can we have a default implementation that doesn't create an empty `ArrayList` and `RecordHeaders` object on every input record?

Reply:
Yeah, thought about that. So it should be a `Collections.emptyList()`?

Reply:
I was thinking about `Optional<Headers> getHeaders(IN input)`, and default is returning `Optional.empty()`.

I'm also wondering if it's possible / makes sense to NOT expose the Kafka `Headers` class on the public API. Although Kafka API seems to be rather stable, in general its best if we can avoid exposing our dependencies' classes on the public interface.

Reply:
Hm... Personally I don't really like that. The big thing about Optionals is to force the handling of possible `null` values. You don't need that protecting while returning collections, as you can always return an empty one, which is what I did. I'd much rather go for an empty (immutable) collection.

Initially, I returned a `Iterable<Header>` and changed that to kafka API Headers upon request. I'm not really leaning to either side, so happy to change that back @RamanVerma, wdyt?

Reply:
@AlexAxeman see my latest comment above about the default case being `headerProvider` is null.

In that case, we no longer need / want to return `Optional`.

i.e.

```
interface HeaderProvider {
    Headers getHeaders(IN input);
}
```

and in `KafkaRecordSerializationSchemaWrapper#serialize()`:
```
if (headerProvider != null) {
    return new ProducerRecord(
      ...,
      headerProvider.getHeaders(input)
    )
} else (
    return new ProducerRecord() // no headers
)
```